bigquery/google/cloud/bigquery/client.py

```diff
@@ -1547,6 +1547,25 @@ def load_table_from_dataframe(
         if location is None:
             location = self.location
 
+        # If table schema is not provided, we try to fetch the existing table
+        # schema, and check if dataframe schema is compatible with it - except
+        # for WRITE_TRUNCATE jobs, the existing schema does not matter then.
+        if (
+            not job_config.schema
+            and job_config.write_disposition != job.WriteDisposition.WRITE_TRUNCATE
+        ):
+            try:
+                table = self.get_table(destination)
+            except google.api_core.exceptions.NotFound:
+                table = None
+            else:
+                # schema fields not present in the dataframe are not needed
+                job_config.schema = [
+                    field
+                    for field in table.schema
+                    if field.name in set(dataframe.columns)
```

Comment:
Let's use `_pandas_helpers.list_columns_and_indexes`, since we support uploading indexes now, and add a test with a named string index to verify that the BigQuery STRING data type is used when a named index matches that of the existing table.

https://github.com/googleapis/google-cloud-python/blob/master/bigquery/google/cloud/bigquery/_pandas_helpers.py#L209

Also, we should make a temporary variable so that we aren't constructing a set each iteration.

```
columns_and_indexes = frozenset(_pandas_helpers.list_columns_and_indexes(dataframe))
job_config.schema = [
    field
    for field in table.schema
    if field.name in columns_and_indexes
]
```